l1-contracts/src/core/messagebridge/MessageBox.sol

```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 Aztec Labs.
+pragma solidity >=0.8.18;
+
+/**
+ * @title MessageBox
+ * @author Aztec Labs
+ * @notice Data structure used in both Inbox and Outbox for keeping track of entries
+ * Implements a multi-set storing the multiplicity (count for easy reading) at the entry.
+ */
+abstract contract MessageBox {
+  error MessageBox__Unauthorized();
+  error MessageBox__NothingToConsume(bytes32 entryKey);
+  error MessageBox__OversizedContent();
+
+  uint256 public constant MESSAGE_SIZE = 32;
+
+  /// @dev Message sender on L1. ChainID if multiple L1s tap into the same Aztec instance.
+  struct L1Actor {
+    address actor;
+    uint256 chainId;
+  }
+
+  /// @dev Message receiver on L2.
+  struct L2Actor {
+    bytes32 actor;
+    uint256 version;
+  }
+
+  /**
+   * @dev Entry struct - Done as struct to easily support extensions if needed
+   * @param count - The occurrence of the entry in the dataset
+   */
+  struct Entry {
```

Comment:
To handle the fees somewhat gracefully, you can change entry to hold the fee and count. The number of equal entryKeys when fees and everything is in there should not be so high, so you will not need a full uint -> packing.